core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkEncodedInputStream.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.chunkedencoding;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.auth.aws.internal.io.SdkInputStream;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+
+
+/**
+ * An implementation of chunked transfer encoding, which wrap an @{link InputStream}.
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ */
+@SdkProtectedApi
+public final class ChunkEncodedInputStream extends SdkInputStream {
+    private static final Logger LOG = Logger.loggerFor(ChunkEncodedInputStream.class);
+    private static final byte[] CRLF = {'\r', '\n'};
+    private static final byte[] END = {};
+
+    private final InputStream inputStream;
+    private final int chunkSize;
+
+    private final ChunkHeader header;
+    private final List<ChunkExtension> extensions = new ArrayList<>();
+    private final List<Trailer> trailers = new ArrayList<>();
+
+    private Chunk currentChunk;
+    private boolean isFinished = false;
+
+    public ChunkEncodedInputStream(
+        InputStream inputStream,
+        int chunkSize,
+        ChunkHeader header,
+        List<ChunkExtension> extensions,
+        List<Trailer> trailers
+    ) {
+        this.inputStream = inputStream;
+        this.chunkSize = chunkSize;
+        this.header = header;
+        this.extensions.addAll(extensions);
+        this.trailers.addAll(trailers);
+    }
+
+    @Override
+    protected InputStream getWrappedInputStream() {
+        return inputStream;
+    }
+
+    public int getChunkSize() {
+        return chunkSize;
+    }
+
+    @Override
+    public int read() throws IOException {
```

Comment:
Don't we need to implement `read(byte b[], int off, int len)` as well? I'm not sure how it would work with the current Chunk implementation though.

Reply:
No, that method is implemented in `InputStream`.